src/desktop/usb.rs

```diff
@@ -0,0 +1,282 @@
+use zbus::zvariant::{
+    DeserializeDict, ObjectPath, OwnedFd, OwnedObjectPath, OwnedValue, SerializeDict, Type, Value,
+};
+
+use crate::{
+    desktop::{HandleToken, Session, SessionPortal},
+    proxy::Proxy,
+    Error, WindowIdentifier,
+};
+
+#[derive(Debug, SerializeDict, Type, Default)]
+#[zvariant(signature = "dict")]
+struct CreateSessionOptions {
+    handle_token: HandleToken,
+    session_handle_token: HandleToken,
+}
+
+/// Options for the USB portal.
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+struct UsbEnumerateOptions {}
+
+/// USB device description
+#[derive(SerializeDict, DeserializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct UsbDevice {
+    parent: Option<String>,
+    /// Device can be opened for reading. Default is false.
+    readable: Option<bool>,
+    /// Device can be opened for writing. Default is false.
+    pub writable: Option<bool>,
+    /// The device node for the USB.
+    #[zvariant(rename = "device-file")]
+    pub device_file: Option<String>,
+    /// Device properties
+    pub properties: Option<HashMap<String, OwnedValue>>,
+}
+
+impl UsbDevice {
+    /// Return the vendor string for display.
+    pub fn vendor(&self) -> Option<String> {
+        self.properties.as_ref().and_then(|properties| {
+            properties
+                .get("ID_VENDOR_FROM_DATABASE")
+                .or_else(|| properties.get("ID_VENDOR_ENC"))
+                .and_then(|v| v.downcast_ref::<String>().ok())
+        })
+    }
+
+    /// Return the model string for display.
+    pub fn model(&self) -> Option<String> {
+        self.properties.as_ref().and_then(|properties| {
+            properties
+                .get("ID_MODEL_FROM_DATABASE")
+                .or_else(|| properties.get("ID_MODEL_ENC"))
+                .and_then(|v| v.downcast_ref::<String>().ok())
+        })
+    }
+}
+
+/// USB error for acquiring device.
+#[derive(Debug)]
+pub struct UsbError(pub Option<String>);
+
+impl From<AcquiredDevice> for Result<OwnedFd, UsbError> {
+    fn from(v: AcquiredDevice) -> Result<OwnedFd, UsbError> {
+        if let Some(fd) = v.fd {
+            if v.success {
+                Ok(fd)
+            } else {
+                Err(UsbError(v.error))
+            }
+        } else {
+            Err(UsbError(None))
+        }
+    }
+}
+
+/// Device to acquire.
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct AcquireDevice {
+    writable: bool,
+}
+
+/// Device to acquire
+pub struct Device(pub String /* ID */, pub bool /* writable */);
+
+/// Option for AcquireDevice call.
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+struct AcquireOptions {
+    handle_token: HandleToken,
+}
+
+/// Finished device acquired.
+#[derive(DeserializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+struct AcquiredDevice {
+    success: bool,
+    fd: Option<OwnedFd>,
```

Comment:
So if success = true, do we always receive an fd or not? or if the success = true and fd = None, there are no more fds to acquire?